pkg/cmd/cmd_add.go

```
@@ -99,17 +107,92 @@ func onlyOnceOption(c *cli.Context, name string) (*string, error) {
+	// parse from 'kpm add -git https://xxx/xxx.git -tag v0.0.1'.
+	if c.NArg() == 0 {
+		gitOpts, err := parseGitRegistryOptions(c)
+		if err != nil {
+			return nil, err
+		}
+		return &opt.AddOptions{
+			LocalPath:    localPath,
+			RegistryOpts: *gitOpts,
+		}, nil
+	} else {
+		// parse from 'kpm add xxx:0.0.1'.
+		ociReg, err := parseOciRegistryOptions(c)
+		if err != nil {
+			return nil, err
+		}
+		return &opt.AddOptions{
+			LocalPath:    localPath,
+			RegistryOpts: *ociReg,
+		}, nil
+	}
+}
+
+// parseGitRegistryOptions will parse the git registry information from user cli inputs.
+func parseGitRegistryOptions(c *cli.Context) (*opt.RegistryOptions, error) {
+	gitUrl, err := onlyOnceOption(c, "git")
+
+	if err != nil {
+		return nil, nil
+	}
+
+	gitTag, err := onlyOnceOption(c, "tag")
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &opt.RegistryOptions{
+		Git: &opt.GitOptions{
+			Url: *gitUrl,
+			Tag: *gitTag,
+		},
+	}, nil
+}
+
+// parseOciRegistryOptions will parse the oci registry information from user cli inputs.
+func parseOciRegistryOptions(c *cli.Context) (*opt.RegistryOptions, error) {
+	ociPkgRef := c.Args().First()
+	name, version := parseOciPkgNameAndVersion(ociPkgRef)
+	if len(version) == 0 {
+		reporter.Report("kpm: default version 'latest' of the package will be downloaded.")
+		version = opt.DEFAULT_OCI_TAG
+	}
+
+	settings, err := settings.GetSettings()
+	if err != nil {
+		return nil, err
+	}
+
+	return &opt.RegistryOptions{
+		Oci: &opt.OciOptions{
+			Reg:     settings.DefaultOciRegistry(),
+			Repo:    settings.DefaultOciRepo(),
+			PkgName: name,
+			Tag:     version,
+		},
+	}, nil
+}
+
+// parseOciPkgNameAndVersion will parse package name and version
+// from string "<pkg_name>:<pkg_version>".
+func parseOciPkgNameAndVersion(s string) (string, string) {
+	parts := strings.Split(s, ":")
+	if len(parts) == 1 {
+		return parts[0], ""
+	}
+
+	if len(parts) > 2 {
+		return "", ""
```

Do we need to return an err here?

---

fixed.